### ISSUE TITLE
no-extra-parens does not allow conditional object spread

### ISSUE THREAD

**Issue by dackmin** [eslint/eslint#13409]
**What rule do you want to change?**
`no-extra-parens`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
The best approach for this might be a new option, because the actual code without the parentheses still works, but can quickly become a real mess.

**Please provide some example code that this change will affect:**

```js
const obj = {
  ...(true ? { foo: 'bar' } : { bar: 'foo' }),
};
```

**What does the rule currently do for this code?**
Shows an error when wrapping a conditional object spread with parentheses.

**What will the rule do after it's changed?**
Allow extra parentheses around conditional object spread.

**Are you willing to submit a pull request to implement this change?**
I'm not familiar with the eslint parser **at all**, so if someone is better at this please be my guest, otherwise I'll try to learn new things ✨

**Comment by kaicataldo** [comment]
Verified using [our demo](https://eslint.org/demo#eyJ0ZXh0IjoiY29uc3Qgb2JqID0ge1xuICAuLi4odHJ1ZSA/IHsgZm9vOiAnYmFyJyB9IDogeyBiYXI6ICdmb28nIH0pLFxufTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLWV4dHJhLXBhcmVucyI6Mn0sImVudiI6e319fQ==).

**Comment by anikethsaha** [comment]
I can work on this if no one is taking this.

**Comment by kaicataldo** [comment]
@anikethsaha Go for it!

**Comment by mdjermanovic** [comment]
Since this isn't a bug, is it accepted as a new option?

**Comment by anikethsaha** [comment]
yeah, probably `enforceForSpreadOperator` ?, default is `true` ?

**Comment by kaicataldo** [comment]
Sorry, I thought the lack of parentheses changed the execution of this code, but I see I was mistaken. Since this is a style preference, I think this falls under our [new rule change policy](https://eslint.org/blog/2020/05/changes-to-rules-policies).

**Comment by dackmin** [comment]
Yep the code example works with or without parentheses, sorry if that wasn't clear.

Here is a _messy_ example, for the record:

```javascript
{
  input: './src/index.js',
  ...format === 'esm' ? {
    manualChunks: id =>
      id.includes('node_modules') ? 'vendor' : path.parse(id).name,
  } : {},
}
```

And I can get messier if needed!

I have yet to find the point where the lack of parentheses for a conditional spread brings any unintended code behavior.

So from what I understand from the change policy, if no one is able to reproduce a case of unintended code behavior regarding this, you can consider it a pure stylistic change and not a syntax issue.

**Comment by kaicataldo** [comment]
I agree that wrapping this expression in parentheses makes the intent a lot clearer. I work on this stuff every day and I wasn't sure of the actual behavior! But I do think this falls under the umbrella of "style preferences", since it's not 100% necessary.

**Comment by anikethsaha** [comment]
Yes, Its a styling change in a `non-stylistic` rule. 
may come under [this](https://eslint.org/blog/2020/05/changes-to-rules-policies) as mentioned above

**Comment by dackmin** [comment]
I'll keep this issue open while you decide whether or not to add this as a new option, feel free to close it if needed 😄 

For those having the same problem, I ended up making a custom plugin overriding `no-extra-parens` and allowing them with a `SpreadElement` (or the legacy `ExperimentalSpreadProperty`) parent:

```js
const eslint = require('eslint');
const ruleComposer = require('eslint-rule-composer');

const rule = new eslint.Linter().getRules().get('no-extra-parens');

module.exports = ruleComposer.filterReports(
  rule,
  problem => {
    if (
      problem.node.type === 'ConditionalExpression' &&
      (
        problem.node.parent.type === 'SpreadElement' ||
        problem.node.parent.type === 'ExperimentalSpreadProperty'
      )
    ) {
      return false;
    }

    return problem;
  },
);
```

**Comment by kaicataldo** [comment]
@dackmin Thanks for sharing! ESLint has always been intended to be used with custom rules, and I'm sure others will appreciate this solution as well!
